File: algorithms/148-sort-list/sort-list.go
package problem148


type ListNode struct {
    Val int
    Next *ListNode
}

// 自顶向下归并排序
func sortList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	// cut the linked list at the mid index
	slow, fast := head, head
	for fast.Next != nil && fast.Next.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	// cut the linkage
    mid := slow.Next
    slow.Next = nil
	left, right := sortList(head), sortList(mid)	
	return merge(left, right)
}

// 自底向上归并排序
func sortListV2(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	length := 0
	node := head
	for node != nil {
		length++
		node = node.Next
	}

	dummyHead := &ListNode{}
	dummyHead.Next = head
	subLen := 1

	for subLen < length {
		prev := dummyHead
		cur := prev.Next

		for cur != nil {
			// first sub list
			head1 := cur
			i := 1
			for i < subLen && cur.Next != nil {
				cur = cur.Next
				i++
			}

			// second sub list
			head2 := cur.Next
			cur.Next = nil // cur the linkage
			i = 1
			cur = head2
			for i < subLen && cur != nil && cur.Next != nil {
				cur = cur.Next
				i++
			}

			// save the next head
			var nextHead *ListNode
			if cur != nil {
				nextHead = cur.Next
				cur.Next = nil
			}

			// merge and reconnect
			prev.Next = merge(head1, head2)
			for prev.Next != nil {
				prev = prev.Next
			}
			
			cur = nextHead
		}

		subLen <<= 1
	}

	return dummyHead.Next
}

func merge(head1 *ListNode, head2 *ListNode) *ListNode {
	dummy_h := &ListNode{}
	h := dummy_h
	for head1 != nil && head2 != nil {
		if head1.Val < head2.Val {
			h.Next = head1
			head1 = head1.Next
		} else {
			h.Next = head2
			head2 = head2.Next
		}
		h = h.Next
	}
	if head1 != nil {
		h.Next = head1
	} else if head2 != nil {
		h.Next = head2
	}

	return dummy_h.Next
}

File: algorithms/148-sort-list/sort-list.py
"""
148. sort list

https://leetcode.com/problems/sort-list/
https://leetcode-cn.com/problems/sort-list/

排序链表
"""


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next: ListNode = None


class Solution:
    """
    自顶向下归并排序

    递归：找到链表的中点，切分，排序，合并

    complexity analysis
    time complexity: O(nlogn)
    space complexity: O(logn) （递归所需的栈空间）
    """

    def sort_list(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        # cut the linked list at the mid index
        # 找到链表中点，可使用快慢双指针
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        # cut the linkage
        mid, slow.next = slow.next, None
        # recursive sort
        left, right = self.sort_list(head), self.sort_list(mid)
        return self.merge(left, right)

    @staticmethod
    def merge(head1: ListNode, head2: ListNode) -> ListNode:
        # merge head1 and head2 linked list and return it
        h = dummy_res = ListNode(0)
        while head1 and head2:
            if head1.val < head2.val:
                h.next = head1
                head1 = head1.next
            else:
                h.next = head2
                head2 = head2.next
            h = h.next

        if head1:
            h.next = head1
        elif head2:
            h.next = head2

        return dummy_res.next


class Solution2:
    """
    自底向上归并排序

    complexity analysis
    time complexity: O(nlogn)
    space complexity: O(1)
    """

    def sort_list(self, head: ListNode) -> ListNode:
        if not head:
            return head
        length = 0
        node = head
        while node:
            length += 1
            node = node.next
        dummy_head = ListNode(0)
        dummy_head.next = head
        sub_length = 1
        while sub_length < length:
            prev = dummy_head
            cur = prev.next

            while cur is not None:
                # 第一个 sub list
                head1 = cur
                i = 1
                while i < sub_length and cur.next is not None:
                    cur = cur.next
                    i += 1

                # 第二个 sub list
                head2 = cur.next
                cur.next = None  # must cut the linkage
                i = 1
                cur = head2
                while i < sub_length and cur is not None and cur.next is not None:
                    cur = cur.next
                    i += 1

                # save the next head
                next_head = None
                if cur is not None:
                    next_head = cur.next
                    cur.next = None

                # merge and reconnect
                prev.next = self.merge(head1, head2)
                while prev.next is not None:
                    prev = prev.next

                cur = next_head

            # sub length * 2
            sub_length <<= 1

        return dummy_head.next

    # noinspection DuplicatedCode
    @staticmethod
    def merge(head1: ListNode, head2: ListNode) -> ListNode:
        # merge head1 and head2 linked list and return it
        h = dummy_res = ListNode(0)
        while head1 and head2:
            if head1.val < head2.val:
                h.next = head1
                head1 = head1.next
            else:
                h.next = head2
                head2 = head2.next
            h = h.next

        if head1:
            h.next = head1
        elif head2:
            h.next = head2

        return dummy_res.next
